### PR TITLE
esp32/heltec_wifi_lora_32: Add HELTEC ESP32 WiFi LoRa 32

### PR DESCRIPTION
**Pull request by adalava**
Product page: https://heltec.org/project/wifi-lora-32/

Supported:
    - LED
    - PRG Button
    - OLED Display
 
Unsupported:
    - LoRa

Tested with:
    - samples\hello_world
    - samples\basic\blinky
    - samples\basic\button
    - samples\subsys\display\lvgl

Signed-off-by: Alfredo Dal'Ava Junior <alfredo.junior@eldorado.org.br>

NOTES:
 - Board code is identical to boards/xtensa/esp32
 - Use 'west build -p auto -b heltec_wifi_lora32_v2 samples/subsys/display/lvgl -- -DSHIELD=ssd1306_128x64' do test the OLED display

### CODE REVIEW

File: boards/xtensa/heltec_wifi_lora32/Kconfig.board [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# HELTEC ESP32 board configuration
+
+# Copyright (c) 2017 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_HELTEC_WIFI_LORA32
+	bool "HELTEC WiFi LoRa 32 (V2) Board"
+	depends on SOC_ESP32
```
Comment (uLipe) [inline boards/xtensa/heltec_wifi_lora32/Kconfig.board:8]: Missing newline

File: boards/xtensa/heltec_wifi_lora32/board.cmake [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+
+include(${ZEPHYR_BASE}/boards/common/esp32.board.cmake)
```
Comment (uLipe) [inline boards/xtensa/heltec_wifi_lora32/board.cmake:3]: missing newline

File: boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+arch: xtensa
+toolchain:
+  - espressif
+supported:
+  - gpio
+  - i2c
+  - watchdog
+  - uart
+  - pinmux
+  - nvs
```
Comment (uLipe) [inline boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32.yaml:13]: missing newline

File: boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+CONFIG_GEN_ISR_TABLES=y
+CONFIG_GEN_IRQ_VECTOR_TABLE=n
+
+CONFIG_I2C=y
+CONFIG_I2C_ESP32=y
+CONFIG_I2C_0=y
+CONFIG_I2C_1=y
+CONFIG_CLOCK_CONTROL=y
+
+CONFIG_BOOTLOADER_ESP_IDF=y
```
Comment (uLipe) [inline boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig:34]: missing newline

**Comment by uLipe** [comment]
Some minor comments, otherwise generally LGTM

File: boards/xtensa/heltec_wifi_lora32/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_BOOTLOADER_ESP_IDF)
```
Comment (shubhamkulkarni97) [inline boards/xtensa/heltec_wifi_lora32/CMakeLists.txt:3]: Is it necessary to duplicate this file? Maybe we can create a common CMake file which can be used by all ESP32 boards.

File: boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_XTENSA_RESET_VECTOR=n
+
+CONFIG_BOARD_HELTEC_WIFI_LORA32=y
+CONFIG_SOC_ESP32=y
```
Comment (shubhamkulkarni97) [inline boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig:6]: Default configs also seem to be same for all ESP32 boards. We can move it to a common file?

**Review by shubhamkulkarni97 — COMMENTED** [review]
@adalava It will be better to maintain the code if we move ESP32 specific code to a common place. Minor comments, otherwise LGTM.

File: boards/xtensa/heltec_wifi_lora32/Kconfig.board (hunk 1 above)
Comment (adalava) [reply boards/xtensa/heltec_wifi_lora32/Kconfig.board]: done, thanks!

File: boards/xtensa/heltec_wifi_lora32/CMakeLists.txt (hunk 5 above)
Comment (adalava) [reply boards/xtensa/heltec_wifi_lora32/CMakeLists.txt]: @shubhamkulkarni97 you're correct. I copied it from the esp32 directory but I verified there's no problem removing it

File: boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig (hunk 6 above)
Comment (adalava) [reply boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig]: It's a good idea. I looked at other boards but I couldn't figure out where to put a common file and make both boards/xtensa/esp32 and boards/xtensa/heltec_wifi_lora32_v2 use it. Do you have a recommendation?

Comment (adalava) [reply boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig]: It looks like @sylvioalves is working on some thing like this. Should we keep this PR as is and move the common config on another PR? (The same for CMake files)

**Comment by adalava** [comment]
Thank you for reviewing it.
I updated the branch to be in sync with main again and added documentation to docs folder, based on the original esp32 port documentation. If something else is needed, just let me know.  I'm also online on slack at #esp32.

File: boards/xtensa/heltec_wifi_lora32_v2/Kconfig.board [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# HELTEC ESP32 board configuration
+
+# Copyright (c) 2017 Intel Corporation
```
Comment (sylvioalves) [inline boards/xtensa/heltec_wifi_lora32_v2/Kconfig.board:3]: I think you can update this copyright to 2021. Eventually change to the company you work for. If none, ok.

File: boards/xtensa/heltec_wifi_lora32_v2/doc/index.rst [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+Build Environment Setup
+=======================
+
+Some variables must be exported into the environment prior to building this port.
+Find more information at :ref:`env_vars` on how to keep this settings saved in you environment.
+
+.. note::
+
+   In case of manual toolchain installation, set :file:`ESPRESSIF_TOOLCHAIN_PATH` accordingly.
+   Otherwise, set toolchain path as below. If necessary, update the version folder path as in :file:`esp-2020r3-8.4.0`.
```
Comment (sylvioalves) [inline boards/xtensa/heltec_wifi_lora32_v2/doc/index.rst:59]: Can you update the guidelines in this file? Toolchain instructions was changed.

File: boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+};
+
+&uart2 {
+	current-speed = <115200>;
+	tx-pin = <17>;
+	rx-pin = <16>;
+	rts-pin = <7>;
+	cts-pin = <8>;
+	hw-flow-control;
+};
```
Comment (sylvioalves) [inline boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:79]: From my perspective, you could remove this unused uart instances.

**Review by sylvioalves — DISMISSED** [review]
Only a few not blocking comments.

**Comment by adalava** [comment]
Thanks for the review, @sylvioalves. I like all your points, the diff was updated and rebased.

**Review by dcpleung — DISMISSED** [review]

File: boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig (hunk 6 above)
Comment (sylvioalves) [reply boards/xtensa/heltec_wifi_lora32/heltec_wifi_lora32_defconfig]: Let's keep as is. Once we refactor board folder structure we re-work this properly.

File: boards/xtensa/heltec_wifi_lora32_v2/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+  board_finalize_runner_args(esp32 "--esp-flash-partition_table=${espidf_build_dir}/partitions_singleapp.bin")
+
+  board_finalize_runner_args(esp32 "--esp-boot-address=0x1000")
+
+  board_finalize_runner_args(esp32 "--esp-partition-table-address=0x8000")
+
+  board_finalize_runner_args(esp32 "--esp-app-address=0x10000")
+
+endif()
```
Comment (sylvioalves) [inline boards/xtensa/heltec_wifi_lora32_v2/CMakeLists.txt:65]: Please, delete this file, it was moved into soc folder.

File: boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+&i2c0 {
+	status = "okay";
+	clock-frequency = <I2C_BITRATE_FAST>;
+	sda-pin = <21>;
+	scl-pin = <22>;
+};
+
+&flash0 {
```
Comment (sylvioalves) [inline boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:74]: You might want to update this if MCUBoot is needed.

File: boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+&cpu1 {
+	clock-frequency = <ESP32_CLK_CPU_240M>;
+};
+
+&uart0 {
+	status = "okay";
+	current-speed = <115200>;
+	tx-pin = <1>;
+	rx-pin = <3>;
+	rts-pin = <22>;
```
Comment (sylvioalves) [inline boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:59]: Remove RTS and CTS pin if not used.

**Comment by sylvioalves** [comment]
@adalava Also check the commit message, CI is failing due to that.

**Comment by adalava** [comment]
> @adalava Also check the commit message, CI is failing due to that.

Thanks! I updated the branch. I hope it's ready now.

File: boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2_defconfig [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+CONFIG_GPIO=y
+CONFIG_GPIO_ESP32=y
+
+CONFIG_GEN_ISR_TABLES=y
+CONFIG_GEN_IRQ_VECTOR_TABLE=n
+
+CONFIG_I2C=y
+CONFIG_CLOCK_CONTROL=y
+
+CONFIG_BOOTLOADER_ESP_IDF=y
```
Comment (sylvioalves) [inline boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2_defconfig:30]: This can be removed as ESP_IDF bootloader is the default.

**Comment by adalava** [comment]
Update: Added OLED support out of the box (now possible thanks to @sylvioalves fixes and tips)

@mbolivar-nordic, I described the "vext control" and "oled reset" pins on the DTS under leds/gpio-leds. Do you have any suggestion of better place to describe it? Thanks!

**Review by uLipe — DISMISSED** [review]

File: boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+/ {
+	model = "heltec_wifi_lora32";
+	compatible = "espressif,esp32";
+
+	aliases {
+		uart-0 = &uart0;
+		i2c-0 = &i2c0;
+		led0 = &led0;
+		sw0 = &button0;
+		vext = &vext;
```
Comment (mbolivar-nordic) [inline boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:19]: Do you really need this alias? Why not use `DT_NODELABEL(vext)` to get the node identifier?

**Review by mbolivar-nordic — COMMENTED** [review]
> @mbolivar-nordic, I described the "vext control" and "oled reset" pins on the DTS under leds/gpio-leds. Do you have any suggestion of better place to describe it? Thanks!

Are they actually LEDs? Or are tehy something else?

File: boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts (hunk 14 above)
Comment (adalava) [reply boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:19]: Works with DT_NODELABEL. Thanks!

Comment (adalava) [reply boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts]: > > @mbolivar-nordic, I described the "vext control" and "oled reset" pins on the DTS under leds/gpio-leds. Do you have any suggestion of better place to describe it? Thanks!
> 
> Are they actually LEDs? Or are tehy something else?

They dont describe any LED. The "vext control" controls a mosftet that turns on/off  the power supply for devices external to the microcontroler  (like the onboard OLED display is included).
The "oled reset" pin is the RST pin of the OLED display.

**Comment by adalava** [comment]
Applied @mbolivar-nordic suggestions, rebased and retested the examples.

**Comment by sylvioalves** [comment]
@adalava, please, rebase.

**Review by uLipe — APPROVED** [review]

**Review by sylvioalves — APPROVED** [review]

**Pull request merged**